Updated MSTG-CODE-3 (#1930)
OWASP/owasp-mstg: Document/0x06i-Testing-Code-Quality-and-Build-Settings.md

As a good practice, as little explanatory information as possible should be provided with a compiled binary. Presence of this additional metadata in the form of debug symbols provides valuable information about the code, for instance function names leaking information about what possibly a function do. Importantly, this metadata is not required to execute a binary and thus it is safe to discard it while building the final release binary, using proper compiler configurations.

When debug symbols are placed in a separate dSYM file, these symbols do not make to the final binary delivered to an end user. While the debug symbols embedded in the compiled binary, they can end up being present in the final binary and thus should be verified before the release.

### Static Analysis

> Use `objdump --syms` to inspect the main binary and other included dylibs, and look for symbols with 'd' (debug) flag associated with them. Check [objdump man page](https://www.unix.com/man-page/osx/1/objdump/ "objdump man page") for information about various other symbol's flag characters. A typical output for a binary containing debug symbols looks like following:

Reviewer: Since we go to the verification step here I introduce the paragraph accordingly and also included info from below. I hope this simplifies the section a bit, what do you think?

```suggestion
To verify the existence of debug symbols you can use objdump from [binutils](https://www.gnu.org/s/binutils/ "Binutils") or [llvm-objdump](https://llvm.org/docs/CommandGuide/llvm-objdump.html "llvm-objdump") to inspect all of the app binaries.

In the following snippet we run objdump over `TargetApp` (the iOS main app executable) to show the typical output of a binary containing debug symbols which are marked with the `d` (debug) flag. Check the [objdump man page](https://www.unix.com/man-page/osx/1/objdump/ "objdump man page") for information about various other symbol flag characters.
```

```bash
$ objdump --syms TargetApp

0000000100007dc8 l    d  *UND* -[ViewController handleSubmitButton:]
000000010000809c l    d  *UND* -[ViewController touchesBegan:withEvent:]
0000000100008158 l    d  *UND* -[ViewController viewDidLoad]
...
000000010000916c l    d  *UND* _disable_gdb
00000001000091d8 l    d  *UND* _detect_injected_dylds
00000001000092a4 l    d  *UND* _isDebugged
...
```

> It is important to note, a binary without debug symbols will not have symbols with 'd' flag associated.
> 

Reviewer: rm: this is clear from the prev. paragraph
```suggestion
```

> `Objdump` from [binutils](https://www.gnu.org/s/binutils/ "Binutils") or [llvm-objdump](https://llvm.org/docs/CommandGuide/llvm-objdump.html "llvm-objdump") can be used. The output from the two may have subtle differences, but output from both will contain flag characters for symbols.

Reviewer: rm: I've moved the info above
```suggestion
```

> To remove the embedded debugging symbols, ensure debug symbols stripping is enabled in Xcode configuration. To strip debugging symbols, set `Strip Debug Symbols During Copy` to `YES` via the project's build settings. Stripping debugging symbols will not only reduce the size of the binary and but also increase the difficulty of reverse engineering.

Reviewer: ```suggestion
To prevent the inclusion of debug symbols, set `Strip Debug Symbols During Copy` to `YES` via the XCode project's build settings. Stripping debugging symbols will not only reduce the size of the binary and but also increase the difficulty of reverse engineering.
```
